tutorials/docs-17-implementing-samplers/index.qmd

```diff
@@ -0,0 +1,479 @@
+
+All in all, that means that the signature for our `AbstractMCMC.step` is going to be the following:
+
+```{julia}
+#| eval: false
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    # `LogDensityModel` so we know we're working with LogDensityProblems.jl model.
+    model::AbstractMCMC.LogDensityModel,
+    # Our sampler.
+    sampler::MALA,
+    # Our sampler state.
+    state::MALAState;
+    kwargs...
+)
+```
+
+Great! Now let's actually implement the full `AbstractMCMC.step` for our `MALA`.
+
+Let's remind ourselves what we're going to do:
+
+1.  Sample a new momentum $p$.
+2.  Compute the log-density of the extended target $\log \bar{\gamma}(x, p)$.
+3.  Take a single leapfrog step $(\tilde{x}, \tilde{p}) = L_{\epsilon}(x, p)$.
+4.  Accept or reject the proposed $(\tilde{x}, \tilde{p})$.
+
+All in all, this results in the following:
+
+```{julia}
+using Random: Random
+using Distributions  # so we get the `MvNormal`
+
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    model_wrapper::AbstractMCMC.LogDensityModel,
+    sampler::MALA,
+    state::MALAState;
+    kwargs...
+)
+    # Extract the wrapped model which implements LogDensityProblems.jl.
+    model = model_wrapper.logdensity
+    # Let's just extract the sampler parameters to make our lives easier.
+    ϵ = sampler.ϵ_init
+    M = sampler.M_init
+    # Extract the current parameters.
+    x = state.x
+    # Sample the momentum.
+    p_dist = MvNormal(zeros(LogDensityProblems.dimension(model)), M)
+    p = rand(rng, p_dist)
+    # Propose using a single leapfrog step.
+    x̃, p̃ = leapfrog_step(model, x, p, ϵ, M)
+    # Accept or reject proposal.
+    logp = LogDensityProblems.logdensity(model, x) + logpdf(p_dist, p)
+    logp̃ = LogDensityProblems.logdensity(model, x̃) + logpdf(p_dist, p̃)
+    logα = logp̃ - logp
+    state_new = if log(rand(rng)) < logα
+        # Accept.
+        MALAState(x̃)
+    else
+        # Reject.
+        MALAState(x)
+    end
+    # Return the "sample" and the sampler state.
+    return state_new.x, state_new
+end
+```
+
+Fairly straight-forward.
+
+Of course, we haven't defined the `leapfrog_step` method yet, so let's do that:
+
+```{julia}
+function leapfrog_step(model, x, p, ϵ, M)
+    # Update momentum `p` using "position" `x`.
+    ∇logγ_x = last(LogDensityProblems.logdensity_and_gradient(model, x))
+    p1 = p + (ϵ / 2) .* ∇logγ_x
+    # Update the "position" `x` using momentum `p1`.
+    x̃ = x + ϵ .* (M \ p1)
+    # Update momentum `p1` using position `x̃`
+    ∇logγ_x̃ = last(LogDensityProblems.logdensity_and_gradient(model, x̃))
+    p2 = p + (ϵ / 2) .* ∇logγ_x̃
```

Comment:
Should this be 
```suggestion
    p2 = p1 + (ϵ / 2) .* ∇logγ_x̃
```
? I think that's what it is in the equations at the start of the page. I know nothing about these sampling algorithms though, so correct me if I'm wrong.